Move to a more explicit memory sharing model

sboeuf opened the pull request:
In anticipation for the huge pages support, the way to share/unshare memory from a TVM is updated with a more explicit model that expects the host to drive the entire operation. Relying on some new TEE-Host functions, the host can invalidate and remove a specific range of memory based on the request from the TVM.

sboeuf commented:
This PR depends on https://github.com/rivosinc/sbi-rs/pull/12 where the extended TEE-Host API is defined.
Also, this requires some additional tests in `tellus`/`guestvm` to validate it's functional, but I posted the PR to get some early feedback.

sboeuf commented:
> Also, this requires some additional tests in `tellus`/`guestvm` to validate it's functional, but I posted the PR to get some early feedback.

I've just updated the tellus VM so that it follows the new model for completing memory sharing and memory unsharing from the guest.

abrestic-rivos commented on src/vm_pages.rs:
```diff
@@ -1264,6 +1319,75 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
+            VmRegionType::ConfidentialRemovable,
+            VmRegionType::Shared,
+            true,
+        )?;
```
need to check that all the pages have been removed first

sboeuf replied:
Ah yes of course, given we've cleared the PTE bits through `remove_pages`, we expect every PTE from the range to be marked as `Unused`.

sboeuf commented on page-tracking/src/page_tracker.rs:
```diff
@@ -338,6 +338,31 @@ impl PageTracker {
+        info.unblock()
+    }
+
+    /// Remove the page, moving it to either "Mapped" or "Converted" state.
+    pub fn remove_page(&self, addr: SupervisorPageAddr) -> Result<()> {
+        let mut page_tracker = self.inner.lock();
+        let info = page_tracker.get_mut(addr).unwrap();
+        info.remove()
+    }
```
If you confirm you're okay replacing the `PageInfo::remove()` behavior with `PageInfo::release()` behavior, then of course I'll replace `remove_page()` with `release_page()`.

sboeuf commented on riscv-page-tables/src/page_table.rs:
```diff
@@ -1046,6 +1053,66 @@ impl<T: PagingMode> GuestStagePageTable<T> {
+                        return Err(Error::PredicateFailed);
+                    }
+                }
+                Unused(_) => continue,
```
Oh so do you also expect that we should be blocking a fully populated range? I'm asking because then I would use the non-sparse version of `invalidate_range()`.

sboeuf commented on src/vm_pages.rs:
```diff
@@ -1751,6 +1752,100 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
+        let invalidated = self
+            .inner
+            .root
+            .invalidate_range_sparse(page_addr, len, |addr| {
```
Got it! That makes sense!

sboeuf replied:
I patched this and then realized the `tellus`/`guestvm` test was broken. Basically what happens is that when the `guestvm` first tries to share some untouched memory, then `tellus` invokes `block_pages()`, but then `salus` only finds `Unused` PTEs from the PT walk.
Do I need to reuse the logic that was present in `share_mem_region_begin()` so that trying to share a range if the page is not mapped would not cause a breaking ecall but instead a resumable one? Or instead, do you think the page should have been mapped by `tellus` beforehand?

abrestic-rivos replied:
ah, in the case that the region is totally unpopulated to begin with, the host does not have to do the block/remove sequence since there's nothing to remove. so in that case the host can just put the return value in A0 and run the vCPU again.

sboeuf replied:
Alright!

sboeuf commented on src/vm_pages.rs:
```diff
@@ -1751,6 +1752,100 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
+        let valid_pages = self
+            .inner
+            .root
+            .validate_range(page_addr, len, |addr| {
+                self.inner.page_tracker.is_blocked_page(
+                    addr,
+                    self.inner.page_owner_id,
+                    MemType::Ram,
+                ) || self
+                    .inner
+                    .page_tracker
+                    .is_blocked_shared_page(addr, MemType::Ram)
+            })
```
No that's the only place where we use it, so I can merge this into a single helper. And I think this will be the same for `promote`/`demote` functions, all they'll care about is if the page is blocked, doesn't matter if it really is `BlockedShared` or `Blocked`.

sboeuf commented on src/vm_pages.rs:
```diff
@@ -1751,6 +1752,100 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
+                .ok_or(Error::AddressOverflow)?,
+        )
+        .ok_or(Error::UnalignedAddress)?;
+        if !self
+            .inner
+            .regions
+            .read()
+            .contains(page_addr, end, VmRegionType::ConfidentialRemovable)
+            && !self
+                .inner
+                .regions
+                .read()
+                .contains(page_addr, end, VmRegionType::SharedRemovable)
+        {
```
Yes I'll improve this.

sboeuf commented on src/vm_pages.rs:
```diff
@@ -1751,6 +1752,100 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
+                    .page_tracker
+                    .is_mapped_page(addr, self.inner.page_owner_id, MemType::Ram)
+                    || (self.inner.page_tracker.is_shared_page(addr, MemType::Ram)
+                        && self.inner.page_tracker.is_owned(addr, PageOwnerId::host()))
```
Out of curiosity, can you elaborate why? Do we expect a shared page to be owned by something else than the host?

sboeuf replied:
Ah I see! thanks for the clarification.

sboeuf commented on src/vm_pages.rs:
```diff
@@ -1751,6 +1752,100 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
+
+        // Does this need to be TLB current_version instead of TLB min_version?
+        let tlb_version = self.inner.tlb_tracker.min_version();
+        // Now check each page is removable before it is unmapped.
+        let unmapped = self
+            .inner
+            .root
+            .unmap_range(page_addr, len, |addr| {
+                self.inner.page_tracker.is_removable_page(
+                    addr,
+                    self.inner.page_owner_id,
+                    MemType::Ram,
+                    tlb_version,
+                )
+            })
+            .map_err(Error::Paging)?;
```
Oh so you suggest I should lock the regions at the top of this function with `let regions = self.inner.regions.read()` so that no update can happen before the unmapping is performed?

sboeuf commented on src/vm.rs:
```diff
@@ -517,6 +519,47 @@ impl<'a, T: GuestStagePagingMode> FinalizedVm<'a, T> {
 
+    // Completes any pending operation from the host for this VM.
+    fn complete_pending_op(&self, active_vcpu: &mut ActiveVmCpu<T>) -> EcallResult<u64> {
+        if let Some(pending_op) = active_vcpu.take_pending_op() {
```
Ok I'll try that!
What's the clean way to retrieve the value of the A0 register with this approach? I was using the `active_vcpu` but with your proposal I won't have access to it when needed.

sboeuf replied:
IIUC, the thing you want to avoid is having A0 being read twice, right?

Could we pass a reference of `FinalizedVm` to `VmCpu::complete_pending_op()`? This way, we would invoke `complete_share_mem_region()` from inside VmCpu.
That being said, I'm not a big fan of this approach because `FinalizedVm` is the superset that contains `VmCpu` and I'm not sure it's clean to pass a reference down to `VmCpu`.

Another approach would be to call `VmCpu::complete_pending_op()` first, and expect it to return the value of the A0 register along with the `pending_op`, so that from `vm.rs`, we have all the information we need to call into `complete_share_mem_region()`. The only thing that is unclear to me with this approach is how we can make the return type of `VmCpu::complete_pending_op()` generic enough so that we wouldn't make it specific to the case we're trying to solve here.

abrestic-rivos replied:
> IIUC, the thing you want to avoid is having A0 being read twice, right?

to be more precise, having the A0 that's being acted upon be different from the one that ends up getting written to the guest's register file. but yes, only reading it once would prevent that.

> That being said, I'm not a big fan of this approach because FinalizedVm is the superset that contains VmCpu and I'm not sure it's clean to pass a reference down to VmCpu.

agreed.

> Actually we can keep the take_pending_op() logic associated with calling VmCpu::complete_pending_op() first. That would simplify what needs to be returned by VmCpu::complete_pending_op() (still need to figure out how to make it generic enough).

i suppose that works too. another thing that comes to mind is using a closure that `complete_pending_op()` invokes on the ecall and the contents of A0. something like:

```
pub fn try_complete_pending_op<F, E>(&mut self, ecall_completion: F) -> Result<(), E>
where
    F: Fn(&SbiMessage, SbiReturnType) -> Result<(), E>
{
    match self.arch.pending_op {
        Some(Mmio(mmio_op)) => {
            // handle this same as before
        }
        Some(Ecall(msg)) => {
            let sbi_ret = ...;
            ecall_completion(&msg, sbi_ret)?;
            self.set_ecall_result(sbi_ret);
        }
        None => (),
    }
    self.arch.pending_op = None;
    Ok(())
}
```

and then call `try_complete_pending_op()` before entering the `active_vcpu.run()` loop in `run_vcpu()`.

sboeuf replied:
Okay I'll implement this way!

sboeuf commented on src/vm_pages.rs:
```diff
@@ -487,6 +487,67 @@ impl VmRegionList {
+            return Err(Error::InsufficientVmRegionSpace);
+        }
+
+        // Now do the split, marking the range to be updated as 'Updating'.
```
Yes I'll remove it.

sboeuf commented on src/vm_pages.rs:
```diff
@@ -1264,6 +1319,75 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
+            false,
+        )?;
+
+        // Should we go further here, by getting the actual list of pages that will be
+        // invalidated, so that we can know for sure if we can return to the guest.
+        Ok(self.inner.tlb_tracker.current_version().increment())
```
I kept incrementing the TLB (following what was already there in `share_mem_region_begin()`) because that's the way to tell the calling function `handle_tee_guest_msg()` that we have a blocking ecall. But I guess with the new design we can always assume a `ShareMemory` or `UnshareMemory` call is a blocking ecall. And that would allow to get rid of this logic here.

sboeuf commented on page-tracking/src/page_info.rs:
```diff
@@ -117,8 +125,15 @@ impl PageInfo {
         match self.state {
-            Converting(_) | Unassigning(_) | Converted | ConvertedLocked | Mapped | VmState
-            | Shared(_) => {
+            Converting(_)
+            | Unassigning(_)
+            | Converted
+            | ConvertedLocked
+            | Mapped
+            | VmState
+            | Shared(_)
+            | Blocked(_)
+            | BlockedShared(_, _) => {
```
Done!

sboeuf commented:
@abrestic-rivos I think I addressed all of your comments.
BTW, are we planning on merging https://github.com/rivosinc/sbi-rs/pull/12? I'm asking because right now this PR relies on my personal fork of `sbi-rs`.

atishp04 commented:
I haven't looked into the PR in details. As per my understanding, this PR will break the existing setup running TVMs as the share memory design is revamped. I was planning to send the RFC with the existing sharing APIs soon. Can we not merge this until host side implementation is also complete ?

The other option is to wait for the sharing APIs to settle down and complete implementation on host/salus side before sending the RFC.

abrestic-rivos commented on page-tracking/src/page_info.rs:
```diff
@@ -164,7 +171,7 @@ impl PageInfo {
             }
             ConvertedLocked => Err(PageTrackingError::PageLocked),
-            Shared(rc) => {
+            Shared(rc) | BlockedShared(rc, _) => {
```
so there's a bit of a wrinkle here i just realized...

we have two ways to end up with a `SharedBlock` page in `release()`:
1. a TVM that had the page blocked was killed before the page was unblocked or removed
2. a TVM that had the shared page mapped (while it was blocked in another TVM) was killed

the behavior here is correct for case (1). for case (2), we want to leave it as `BlockedShared` but decrement the reference count. seems like we need to know whether or not the page was invalidated (indicating it was blocked) when releasing the page in order to differentiate between the two cases.

sboeuf replied:
I see... Should we carry the TVM id through the `BlockedShared()` state? That could help us know what to do. Or did you have another idea about how to do it?

abrestic-rivos replied:
i did think about that, but it grows the `PageInfo` struct by 64-bits. that's a non-trivial amount of overhead considering the struct is repeated for every page of memory in the system.

another option is to have separate, e.g.`release_present()` vs `release_invalidated()`, calls for the two paths.

sboeuf replied:
>another option is to have separate, e.g.release_present() vs release_invalidated(), calls for the two paths.

Ok I'm gonna try this approach

sboeuf replied:
I've added the commit https://github.com/rivosinc/salus/pull/261/commits/f1bdd023ea5d41f82b377909d1724c664ad54e66 to handle this issue, please let me know if this is covering what you had in mind.

abrestic-rivos replied:
yes, that looks right to me. squash with this commit?

abrestic-rivos commented on src/vm_pages.rs:
```diff
@@ -1334,6 +1463,79 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
 
+    /// Converts the specified memory region from shared to confidential. Returns the TLB version
+    /// at which the conversion will be completed.
+    pub fn unshare_mem_region(&self, page_addr: GuestPageAddr, len: u64) -> Result<()> {
```
the `*unshare_mem_region` methods look pretty similar to the `*share_mem_region` ones except the region types are flipped. is it possible to deduplicate these a bit?

abrestic-rivos commented on src/vm.rs:
```diff
@@ -2359,14 +2397,13 @@ impl<'a, T: GuestStagePagingMode> FinalizedVm<'a, T> {
-                    ),
                     Ok(_) => EcallAction::Break(
-                        VmExitCause::ResumableEcall(SbiMessage::TeeGuest(guest_func)),
+                        VmExitCause::BlockingEcall(
+                            SbiMessage::TeeGuest(guest_func),
+                            self.vm_pages().min_tlb_version(),
+                        ),
```
now that the host is responsible for explicitly blocking & removing the pages, we should remove this logic to block until a specific tlb version is met.

abrestic-rivos replied:
i think we can drop the `Blocked` state entirely now that we handle that logic in `complete_pending_op()`.

sboeuf replied:
Sounds good

abrestic-rivos commented on src/vm_pages.rs:
```diff
@@ -304,12 +304,6 @@ enum VmRegionType {
     Imsic,
```
something to consider, not necessarily for this PR, is to adapt the whole IMSIC bind/unbind/rebind bits to internally use the new block/remove flow. that would allow us to completely get rid of the existing "unassignment" flow.

sboeuf commented:
> I haven't looked into the PR in details. As per my understanding, this PR will break the existing setup running TVMs as the share memory design is revamped. I was planning to send the RFC with the existing sharing APIs soon. Are we planning to merge this PR soon ? We need to point to a working salus branch instead of main in the docs in that case.
> 
> The other option is to wait for the sharing APIs to settle down and complete implementation on host/salus side before sending the RFC. That works for me as well as it may be little odd to change the APIs between between RFC and v1 :)

IMO it'd be best to update the RFC to match the new design, otherwise we might confuse the reviewers upstream, and this might increase the delay in getting the RFC agreed on.

abrestic-rivos commented on page-tracking/src/page_info.rs:
```diff
@@ -159,6 +159,24 @@ impl PageInfo {
 
     /// Pops the current owner if there is one, returning the page to the previous owner.
     pub fn release(&mut self) -> PageTrackingResult<()> {
```
nit: i think it'd be slightly cleaner if you had a single release with a `was_invalid: bool` param that was used to determine the behavior in the `BlockedShared` state, rather than having `release()` fall through to `release_invalidated()`.

abrestic-rivos commented on src/vm_pages.rs:
```diff
@@ -1751,6 +1750,93 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
+            .inner
+            .root
+            .invalidate_range(page_addr, len, |addr| {
+                self.inner
+                    .page_tracker
+                    .is_mapped_page(addr, self.inner.page_owner_id, MemType::Ram)
+                    || (self.inner.page_tracker.is_shared_page(addr, MemType::Ram)
+                        && !self
+                            .inner
+                            .page_tracker
+                            .is_owned(addr, self.inner.page_owner_id))
```
kind of pre-existing and not something that needs to be fixed in this PR, but this logic should probably be in a single `PageTracker` call to avoid bouncing of that lock so much.

abrestic-rivos commented on src/vm_pages.rs:
```diff
@@ -1262,6 +1318,151 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
+        self.do_share_mem_region(page_addr, len, false)
+    }
+
+    pub fn do_share_mem_region(
```
nit: maybe `do_convert_*` for these helpers instead of `do_share_*`? though i guess "convert" is overloaded...

sboeuf replied:
Yes that's fine I can use `do_convert_region_*`

abrestic-rivos commented on src/vm_cpu.rs:
```diff
@@ -820,10 +817,10 @@ impl<'vcpu, 'pages, 'host, T: GuestStagePagingMode> ActiveVmCpu<'vcpu, 'pages, '
                 self.report_ecall_exit(msg);
             }
-            ForwardedEcall(msg) => {
+            ForwardedEcall(msg) | BlockingEcall(msg) => {
```
i think we can totally get rid of the `BlockingEcall` exit cause, right?

sboeuf replied:
We use it for setting up the `pending_op`. Should I replace it with `ForwardedEcall` so that I can fully remove it?

sboeuf replied:
As per our offline conversation, I've replaced BlockingEcall with ForwardedEcall.

abrestic-rivos commented on src/vm.rs:
```diff
@@ -533,6 +533,44 @@ impl<'a, T: GuestStagePagingMode> FinalizedVm<'a, T> {
             .map_err(|_| EcallError::Sbi(SbiError::InvalidParam))?;
+
+        // Complete a possible pending operation
+        active_vcpu.try_complete_pending_op(|sbi_msg, sbi_ret| -> EcallResult<()> {
```
maybe hoist the contents of the closure into it's own function?

sboeuf replied:
This will be cleaner yes.

abrestic-rivos reviewed (commented):
just a couple minor comments left, otherwise lgtm -- thanks!

abrestic-rivos commented on src/vm_pages.rs:
```diff
@@ -616,7 +615,7 @@ impl<'a, T: GuestStagePagingMode, M> VmPagesMapper<'a, T, M> {
             .checked_add_pages(num_pages)
             .ok_or(Error::AddressOverflow)?;
         let regions = vm_pages.regions.read();
-        if !regions.contains(page_addr, end, region_type) {
+        if !regions.contains(page_addr, end, |r| r == region_type) {
             return Err(Error::InvalidMapRegion);
         }
         let mapper = vm_pages
```
per our offline conversation, i think we should allow page faults in `*Removable` regions to continue to be serviceable as if they were the original region type. in other words:
- page faults taken in `*Removable` regions should not be fatal
- host should be allowed to add shared pages to `SharedRemovable` regions, and confidential zero pages to `ConfidentialRemovable`.

sboeuf replied:
Sounds good! What about measured pages?

sboeuf replied:
Please see the new commit handling this: https://github.com/rivosinc/salus/pull/261/commits/7c997b79e15b00b9a701a501286145f07b947054

abrestic-rivos replied:
measured pages can only be added prior to tvm finalization, so it's impossible for those to be in a `*Removable` region anyway (which you've handled correctly.)

abrestic-rivos reviewed (approved):
lgtm! @atishp04 / @rajnesh-kanwal how do you want to handle submitting this? do you want to take a stab at getting the kernel side working first?

rajnesh-kanwal commented:
> lgtm! @atishp04 / @rajnesh-kanwal how do you want to handle submitting this? do you want to take a stab at getting the kernel side working first?

The change overall looks good to me. I don't think there will be problems while supporting it in KVM. Just doing a final review and once done I think we can merge this. I will then start to add host side support in KVM.

rajnesh-kanwal commented on test-workloads/src/bin/tellus.rs:
```diff
@@ -954,11 +979,26 @@ extern "C" fn kernel_init(hart_id: u64, fdt_addr: u64) {
                                 panic!("Ran out of pages to fault in");
                             }
+
+                            if unshared_gpa_range
+                                .as_ref()
+                                .filter(|r| r.contains(&addr))
+                                .is_some()
+                                && unshared_spa_range.is_none()
+                            {
+                                unshared_spa_range = Some(Range {
+                                    start: zero_pages_base + zero_pages_added * PAGE_SIZE_4K,
+                                    end: zero_pages_base
+                                        + zero_pages_added * PAGE_SIZE_4K
+                                        + PAGE_SIZE_4K,
+                                });
+                            }
```
Can you just add a small comment for this part explaining why we are saving unshared_spa_range here.
As per my understanding this is so that we can check if we need a fence or not and run the if block on line `810`. Right?
Seemed confusing to me at the start.

sboeuf replied:
Yes you understood that well. Of course I can add a comment to make things clearer.

rajnesh-kanwal reviewed (approved):
One minor comment otherwise LGTM.
